File: s2graph_gremlin/src/main/java/org/apache/s2graph/gremlin/structure/S2VertexProperty.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.s2graph.gremlin.structure;

import org.apache.tinkerpop.gremlin.structure.Property;
import org.apache.tinkerpop.gremlin.structure.Vertex;
import org.apache.tinkerpop.gremlin.structure.VertexProperty;

import java.util.*;

public class S2VertexProperty<V> implements VertexProperty<V> {
    private S2Vertex element;
    private String key;
    private V value;
    private Map<String, Property<?>> props;

    public S2VertexProperty(S2Vertex element, String key, V value) {
        this.element = element;
        this.key = key;
        this.value = value;
        this.props = new HashMap<>();
    }

    @Override
    public String key() {
        return key;
    }

    @Override
    public V value() throws NoSuchElementException {
        return value;
    }

    @Override
    public boolean isPresent() {
        return true;
    }

    @Override
    public Vertex element() {
        return element;
    }

    @Override
    public void remove() {

    }

    @Override
    public Object id() {
        S2VertexId vertexId = element.getVertexId();
        return vertexId.getServiceName() + "|" +
                vertexId.getColumnName() + "|" +
                vertexId.getId() + "|" + key;
    }

    @Override
    public <V> Property<V> property(String s, V v) {
        VertexProperty<V> newProperty = new S2VertexProperty<V>(element, key, v);
        props.put(key, newProperty);
        return newProperty;
    }

    @Override
    public <U> Iterator<Property<U>> properties(String... propertyKeys) {
        List<Property<U>> ls = new ArrayList<>();
        for (String propertyKey : propertyKeys) {
            ls.add(property(propertyKey));
        }
        return ls.iterator();
    }
}

File: s2core/src/main/scala/org/apache/s2graph/core/S2VertexProperty.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.s2graph.core

import java.util

import org.apache.s2graph.core.mysqls.ColumnMeta
import org.apache.s2graph.core.types.CanInnerValLike
import org.apache.tinkerpop.gremlin.structure.{Property, VertexProperty, Vertex => TpVertex}

import scala.util.hashing.MurmurHash3

case class S2VertexPropertyId[V](columnMeta: ColumnMeta, value: V)

case class S2VertexProperty[V](element: S2Vertex,
                               columnMeta: ColumnMeta,
                               key: String,
                               v: V) extends VertexProperty[V] {
  import CanInnerValLike._
  implicit lazy val encodingVer = element.serviceColumn.schemaVersion
  lazy val innerVal = CanInnerValLike.anyToInnerValLike.toInnerVal(value)
  private var removed = false

  val value = castValue(v, columnMeta.dataType).asInstanceOf[V]

  def toBytes: Array[Byte] = {
    innerVal.bytes
  }

  override def properties[U](strings: String*): util.Iterator[Property[U]] = ???

  override def property[V](key: String, value: V): Property[V] = ???

  override def remove(): Unit = removed = true

  override def id(): AnyRef = S2VertexPropertyId(columnMeta, v)

  override def isPresent: Boolean = !removed

  override def hashCode(): Int = {
    (element, id()).hashCode()
  }

  override def equals(other: Any): Boolean = other match {
    case p: VertexProperty[_] =>
      System.identityHashCode(element) == System.identityHashCode(p.element) && id() == p.id()
    case _ => false
  }

  override def toString(): String = {
    Map("columnMeta" -> columnMeta.toString, "key" -> key, "value" -> value).toString
  }
}
